File: src/lua/dynamic/dynamic.lua
local decorators = require('src.lua.utils.decorators')

local INF = math.huge

local function hamiltonian_dp(graph, context)
  local n = #graph
  if n == 0 then
    return nil, INF
  end

  local size = 2 ^ n
  local dp = {}
  for mask = 0, size - 1 do
    dp[mask] = {}
    for i = 1, n do
      dp[mask][i] = INF
    end
  end
  dp[1][1] = 0 -- Empezamos en el nodo 1

  for mask = 1, size - 1 do
    for i = 1, n do
      if (mask & (1 << (i - 1))) ~= 0 then
        for j = 1, n do
          if i ~= j and (mask & (1 << (j - 1))) ~= 0 then
            if context and context._internal_counter then
              context._internal_counter[1] = context._internal_counter[1] + 1
            end

            local prev_mask = mask ~ (1 << (i - 1))
            if dp[prev_mask][j] ~= INF and graph[j][i] ~= INF then
              local new_cost = dp[prev_mask][j] + graph[j][i]
              if new_cost < dp[mask][i] then
                dp[mask][i] = new_cost
              end
            end
          end
        end
      end
    end
  end

  local final_mask = size - 1
  local min_cost = INF
  local last_vertex = -1

  for i = 2, n do
    if dp[final_mask][i] ~= INF and graph[i][1] ~= INF then
      local cost = dp[final_mask][i] + graph[i][1]
      if cost < min_cost then
        min_cost = cost
        last_vertex = i
      end
    end
  end

  if last_vertex == -1 then
    return nil, INF
  end

  -- Reconstruir el camino
  local path = {}
  for i = 1, n + 1 do path[i] = 0 end
  path[n + 1] = 1
  path[1] = 1
  local current_mask = final_mask
  local current_vertex = last_vertex

  for i = n, 2, -1 do
    path[i] = current_vertex
    local prev_mask = current_mask ~ (1 << (current_vertex - 1))
    for j = 1, n do
      if (prev_mask & (1 << (j - 1))) ~= 0 and
          dp[prev_mask][j] ~= INF and
          graph[j][current_vertex] ~= INF then
        local expected = dp[prev_mask][j] + graph[j][current_vertex]
        if math.abs(dp[current_mask][current_vertex] - expected) < 1e-9 then
          current_vertex = j
          current_mask = prev_mask
          break
        end
      end
    end
  end

  return path, min_cost
end

-- Aplicar decoradores en orden: count → memory → time
hamiltonian_dp = decorators.time_it(
  decorators.measure_memory(
    decorators.count_ops(hamiltonian_dp)
  )
)

return hamiltonian_dp

File: src/python/dynamic/dynamic.py
from src.python.utils.decorators import count_dp_ops
from src.python.utils.decorators import measure_memory
from src.python.utils.decorators import time_it

INF = float('inf')


@time_it
@count_dp_ops
@measure_memory
def hamiltonian_dp(graph, counter=None):
  n = len(graph)
  if n == 0:
    return (None, INF)
  dp = [[INF] * n for _ in range(1 << n)]
  dp[1][0] = 0
  for mask in range(1, 1 << n):
    for i in range(n):
      if (mask >> i) & 1:
        for j in range(n):
          if i != j and (mask >> j) & 1:
            if counter:
              counter[0] += 1
            prev_mask = mask ^ (1 << i)
            if dp[prev_mask][j] != INF and graph[j][i] != INF:
              dp[mask][i] = min(dp[mask][i], dp[prev_mask][j] + graph[j][i])
  final_mask = (1 << n) - 1
  min_cost = INF
  last_vertex = -1
  for i in range(1, n):
    if dp[final_mask][i] != INF and graph[i][0] != INF:
      cost = dp[final_mask][i] + graph[i][0]
      if cost < min_cost:
        min_cost = cost
        last_vertex = i
  if last_vertex == -1:
    return (None, INF)
  path = [0] * (n + 1)
  current_mask = final_mask
  current_vertex = last_vertex
  for i in range(n - 1, 0, -1):
    path[i] = current_vertex
    prev_mask = current_mask ^ (1 << current_vertex)
    for j in range(n):
      if (prev_mask >> j
         ) & 1 and dp[prev_mask][j] != INF and graph[j][current_vertex] != INF:
        if abs(dp[current_mask][current_vertex] -
               (dp[prev_mask][j] + graph[j][current_vertex])) < 1e-9:
          current_vertex = j
          current_mask = prev_mask
          break
  path[0] = 0
  path[n] = 0
  return (path, min_cost)
